glusterd: Optmize RCU_READ_(LOCK|UNLOCK) macros

mohit84 opened the pull request:
Change the definition of macros to avoid glusterfs_this_location
call again and again

Fixes: #1619
Change-Id: I4c9d4161d24d60366323ca420aad6a9641057d61
Signed-off-by: Mohit Agrawal <moagrawa@redhat.com>

mohit84 commented:
/run regression

mohit84 commented:
/run brick-mux regression

mohit84 commented:
/run regression

mohit84 commented:
/run brick-mux regression

mohit84 commented:
/run regression

mohit84 commented:
/run brick-mux regression

xhernandez commented on xlators/mgmt/glusterd/src/glusterd.h:
```diff
@@ -764,18 +764,14 @@ typedef ssize_t (*gd_serialize_t)(struct iovec outmsg, void *args);
     } while (0)
 
 #define RCU_READ_LOCK                                                          \
-    pthread_mutex_lock(&(THIS->ctx)->cleanup_lock);                            \
```
Can these locks be removed safely ? I guess they were added for some reason...

mohit84 replied:
This code was introduced by this patch(https://review.gluster.org/#/c/glusterfs/+/21743/) to avoid a glusterd crash at the time of reboot and big_handler call at the same time.I believe the approach was not correct, will send a new patch to handle the cleanup during any actor function calling.

xhernandez replied:
Have you checked how difficult it would be to correctly terminate requests before calling xlator's `fini()` ?

mohit84 commented:
/run regression

mohit84 commented:
/run regression

mohit84 committed:
glusterd: Call RCU_READ_(UN)LOCK without any mutex

In the commit 2bb0e89e4bb113a93c6e786446a140cd99261af8 to avoid
a gluster crash at the time of calling cleanup_and_exit call
RCU_READ(UN)LOCK under cleanup_lock, that was not a correct way
to avoid a crash.Introtoduce a new flag call_fini at
glusterd_conf_t to avoid race condition during handling of
big_lock actor function while cleanup has been called.

Fixes: #1619
Change-Id: I553c3f717980a21a9935dc7c0d64ca8d415ea289
Signed-off-by: Mohit Agrawal <moagrawa@redhat.com>

mohit84 commented:
/run regression

mohit84 commented:
/run brick-mux regression

mohit84 commented:
/run regression

mohit84 commented:
/run brick-mux regression

mohit84 commented:
/run brick-mux regression
